package PS;
import "util/proto/range.proto";
import "data/proto/data.proto";
import "system/proto/node.proto";
import "parameter/proto/param.proto";
import "filter/proto/filter.proto";
import "learner/proto/sgd.proto";
import "learner/proto/bcd.proto";

message Task {
  // true: this is a system control message
  // false: this is a customer message, and customer_id should be set
  optional bool control = 1 [default = false];
  // whether request or reply message
  optional bool request = 2 [default = false];
  // the unique id of a customer
  optional int32 customer_id = 3;

  // time
  optional int32 time = 5;
  repeated int32 wait_time = 6;

  // the key range of this task
  optional PbRange key_range = 7;
  // namespace of keys
  optional int32 key_channel = 8;
  // indiciate whether or not has the key list
  optional bool has_key = 9 [default = false];

  optional DataType key_type = 13;
  repeated DataType value_type = 14;

  repeated FilterConfig filter = 12;

  optional bytes msg = 17;
  optional Control ctrl = 18;

  // parameters
  optional CallSharedPara shared_para = 20;
  optional SGDCall sgd = 21;
  optional BCDCall bcd = 22;

  extensions 100 to 199;
}

message Control {
  enum Command {
    // a node => the scheduler
    REQUEST_APP = 1;
    REGISTER_NODE = 2;
    REPORT_PERF = 3;
    READY_TO_EXIT = 4;

    // the scheduler => a node
    ADD_NODE = 10;
    UPDATE_NODE = 11;
    REPLACE_NODE = 12;
    REMOVE_NODE = 13;
    EXIT = 14;
  }
  required Command cmd = 1;
  repeated Node node = 2;
}

enum DataType {
  OTHER = 0;
  INT8 = 1;
  INT16 = 2;
  INT32 = 3;
  INT64 = 4;
  UINT8 = 5;
  UINT16 = 6;
  UINT32 = 7;
  UINT64 = 8;
  FLOAT = 9;
  DOUBLE = 10;
  CHAR = 11;
}
